### PR TITLE
Added logic to auto detect system level proxy

### PR DESCRIPTION
**Pull request by chinmay-browserstack**
- Detect system level proxy and log a warning message
- added option useSystemProxy to use proxy which was detected

### CODE REVIEW

File: packages/client/src/detect-proxy.js [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+          dictionary[key] = value === '1';
+        } else if (key.endsWith('Port')) {
+          dictionary[key] = parseInt(value);
+        } else {
+          dictionary[key] = value;
+        }
+      }
+    });
+    const proxies = [];
+    for (const t of filter) {
```
Comment (ninadbstack) [inline packages/client/src/detect-proxy.js:27]: dont use single letter var names

File: packages/client/src/detect-proxy.js [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+          dictionary[`${type}Port`]
+        ) {
+          proxies.push({
+            type: type,
+            host: dictionary[`${type}Proxy`],
+            port: dictionary[`${type}Port`]
+          });
+        }
+      }
+      return proxies;
```
Comment (ninadbstack) [inline packages/client/src/detect-proxy.js:44]: Move these to separate functions - reduce per function line length

File: packages/client/src/detect-proxy.js [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+export default class DetectProxy {
+  constructor() {
+    this.execPromise = promisify(exec);
+    this.platform = process.platform;
+    // There are sock proxies as well which we don't need
+    this.filter = ['HTTP', 'HTTPS'];
+  }
+
+  async getSystemProxy() {
```
Comment (ninadbstack) [inline packages/client/src/detect-proxy.js:13]: function is marked async but there is no await in it

File: packages/client/src/detect-proxy.js [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    if (this.platform === 'darwin') {
+      return this.getProxyFromMac();
+    } else if (this.platform === 'win32') {
+      return this.getProxyFromWindows();
+    } else if (this.platform !== 'linux') {
+      logger('client:detect-proxy').debug(`Not able to auto detect system proxy for ${this.platform} platform`);
+    }
+  }
+
+  async getProxyFromMac() {
```
Comment (rishigupta1599) [inline packages/client/src/detect-proxy.js:23]: Can you add example of string which you are splitting to understand the structure

File: packages/client/src/detect-proxy.js [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    this.filter = ['HTTP', 'HTTPS'];
+  }
+
+  async getSystemProxy() {
+    if (this.platform === 'darwin') {
+      return this.getProxyFromMac();
+    } else if (this.platform === 'win32') {
+      return this.getProxyFromWindows();
+    } else if (this.platform !== 'linux') {
+      logger('client:detect-proxy').debug(`Not able to auto detect system proxy for ${this.platform} platform`);
```
[the remark this answers is not in the document]
Comment (chinmay-browserstack) [reply packages/client/src/detect-proxy.js:19]: This is for platforms other than linux

File: packages/client/src/detect-proxy.js (hunk 4 above)
Comment (chinmay-browserstack) [reply packages/client/src/detect-proxy.js:24]: Added for both windows and mac

**Review by rishigupta1599 — APPROVED** [review]

**Review by ninadbstack — APPROVED** [review]

**Pull request merged**